ringpop.thrift-gen

```diff
@@ -0,0 +1,170 @@
+    adapter := &{{ $Adapter }}{
+        impl:    impl,
+        ringpop: rp,
+        ch:      ch,
+        config:  config,
+    }
+    // create ringpop router for routing based on ring membership
+    adapter.router = ringpop.NewRouter(rp, adapter)
+
+    // notify implementation it is adapted
+    if adaptable, ok := impl.(tchan{{ $Service.Name }}Adaptable); ok {
+        adaptable.SetAdapter(adapter)
+    }
+
+    return adapter
+}
+
+// implement ringpop.ClientFactory
+func (t *{{ $Adapter }}) MakeClient(dest string) interface{} {
+    if dest == t.ringpop.WhoAmI() {
+        // use local client
+        return t.impl
+    }
+
+    // create remote client
+    client := thrift.NewClient(
+        t.ch,
+        t.ch.ServiceName(),
+        &thrift.ClientOptions{
+            HostPort: dest,
+        },
+    )
+    return New{{ $TChanInterface }}Client(client)
+}
+
+{{ range .Methods }}
+{{ $Method := .}}
+{{ $PublicName := $Method.Name | goPublicName }}
+{{ $cfg := print "config." $PublicName }}
+{{ $ConfigType := print ($Service.Name | goPublicName) $PublicName "Configuration"}}
+
+// ------ 'Generated' code for {{ $Method.Name | goPublicName }} ------ {{ $cfg }}
+type {{ $ConfigType }} struct {
+    N       ringpop.ReplicationFactor
+    Fanout  ringpop.ReplicationFactor
+    Reducer func({{ if $Method.ReturnType }}rs []{{ $Method.ReturnType | goType }}, {{ end }}errs []error) ({{ template "returnType" $Method}})
+    Key     func({{ template "arguments" $Method}}) (string, error)
+}
+
+func (t *{{ $Adapter }}) {{ $PublicName }} ({{ template "arguments" $Method }}) ({{ template "returnType" $Method }}) {
+    // check if the function is configured to be ring aware
+    if t.{{ $cfg }} == nil {
+        return t.impl.{{ $PublicName }}({{ template "call" $Method }})
+    }
+
+    if isForwarded(ctx) {
+        return t.impl.{{ $PublicName }}({{ template "call" $Method }})
+    }
+
+    // find the key to shard on
+    ringpopKey, err := t.{{ $cfg }}.Key({{ template "call" $Method }})
+    if err != nil {
+        return {{ if $Method.ReturnType }}r, {{ end }}fmt.Errorf("could not get key: %s", err)
+    }
+
+    clients := t.router.GetClients(ringpopKey, t.{{ $cfg }}.N, t.{{ $cfg }}.Fanout)
+    clientLength := len(clients)
+    if clientLength == 0 {
+        return {{ if $Method.ReturnType }}r, {{ end }}errors.New("no clients")
+    }
+
+    // slices to collect return values
+    {{ if $Method.ReturnType }}
+        rs := make([]{{ $Method.ReturnType | goType }}, clientLength)
+    {{ end }}
+    errs := make([]error, clientLength)
+
+    // make headers indicate a forwarded request
+    ctx = makeForwarded(ctx)
+
+    var wg sync.WaitGroup
+    for i, untypedClient := range clients {
+        client := untypedClient.({{ $TChanInterface }})
+        wg.Add(1)
+        go func(i int, client {{ $TChanInterface }}) {
+            defer wg.Done()
+
+            {{ if $Method.ReturnType }}
+                rs[i], errs[i] =
```

Comment:
I will look into how slices work with concurrent access to different indices.

I was assuming it being thread safe since every `go func` is accessing its own entry in the slices. So there is no concurrent access to a given index in the slice, but it might be that te slice as a whole is accessed concurrently.